File: February/feb04/feb04.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def solve(node, diameter):
    if node is None:
        return 0  # Base case: If node is None, height is 0
    
    # Recursively find the height of left and right subtrees
    left_height = solve(node.left, diameter)
    right_height = solve(node.right, diameter)
    
    # Update the maximum diameter found so far
    diameter[0] = max(diameter[0], left_height + right_height)
    
    # Return the height of the current node
    return 1 + max(left_height, right_height)

def diameter(root):
    diameter = [0]  # Variable to store the maximum diameter
    solve(root, diameter)  # Call helper function
    return diameter[0]  # Return the final diameter

File: February/feb04/feb04.java
class Node {
    int data;
    Node left, right;

    public Node(int data) {
        this.data = data;
        left = right = null;
    }
}

public class BinaryTree {
    // Helper function to calculate height and update diameter
    private static int solve(Node node, int[] diameter) {
        if (node == null) return 0; // Base case: If node is NULL, height is 0

        // Recursively find the height of left and right subtrees
        int leftHeight = solve(node.left, diameter);
        int rightHeight = solve(node.right, diameter);

        // Update the maximum diameter found so far
        diameter[0] = Math.max(diameter[0], leftHeight + rightHeight);

        // Return the height of the current node
        return 1 + Math.max(leftHeight, rightHeight);
    }

    public static int diameter(Node root) {
        int[] diameter = new int[1]; // Variable to store the maximum diameter
        solve(root, diameter); // Call helper function
        return diameter[0]; // Return the final diameter
    }
}
